### DISCUSSION TITLE
Mongoid strange ruby behavior

### DISCUSSION THREAD

**Discussion by CharlesPlantePulsar** [mongodb/mongoid#5776] [Q&A] [answered]
I have a strange behavior using Ruby and Mongoid

I have a Client model (simplified)
```
class Client
  include Mongoid::Document
  attr_accessor :full_address
  has_many :factures, foreign_key: :client
  has_many :articles, foreign_key: :client
  has_many :utilisateurs, foreign_key: :client
....
```
 and and Utilisateur (user in french) that belong to one client:
```
class Utilisateur
  include Mongoid::Document
  has_many :articles, foreign_key: :utilisateur
  field :name, type: String
  field :role, type: String
  field :sizes, type: Array
  field :betweenLegs, type: String
  field :inactive, type: Boolean
  field :_forSearchByText, type: String
  field :matricule, type: String
  
  has_many :measurements

  belongs_to :client, foreign_key: :client
...
```

My problem is that when a make a POST/PUT (create/update), even if I set the client manually, the object writting in the database doesnt hold a reference to my client.

Here's my create method :
```
def create
    utilisateur = Utilisateur.new(utilisateur_params)
    utilisateur.client = Client.first() # I try to get the first one just to be sure its found
    utilisateur.save!

    render json: utilisateur, status: 200
  end
```
Note that on me POST result i got something that i would exepect: 
```
{
    "_forSearchByText": null,
    "_id": "65ae5eb1f8bb26b629750208",
    "betweenLegs": "1325",
    "client": "582f6ce7a12a0e1770ee2762",
    "inactive": false,
    "matricule": "aaaaaaa",
    "name": "aaaaaaa",
    "role": null,
    "sizes": null
}
```
Problem is that on my database I still got no client field:
```
{
  "_id": {
    "$oid": "65ae5f28f8bb26b629750209"
  },
  "name": "aaaaaaa",
  "betweenLegs": "1325",
  "matricule": "aaaaaaa",
  "inactive": false
}
```

I found a way to "solve" the problem by changing remove the foreign_key of client
`belongs_to :client`
This make Ruby use `client_id` instead for the name of foreing_key attributes.
```
{
  "_id": {
    "$oid": "65ae60e9f8bb26b62975020a"
  },
  "name": "aaaaaaa",
  "betweenLegs": "1325",
  "matricule": "aaaaaaa",
  "client_id": {
    "$oid": "582f6ce7a12a0e1770ee2762"
  },
  "inactive": false
}
```

Final problem is that my web app is dependent on another backend (NodeJS) who is using "client". Is there someting I could have missed because I did this with other class that look the same without problem.

Any kind of help will be appraciated :)

**Comment by jamis** [comment, marked as answer]
Hey @CharlesPlantePulsar -- this is a tricky one. The problem is that the `belongs_to` macro creates getter methods both for the association, and for the foreign key. In this case, both are named the same thing, so it creates the `client` accessor for the association first, and then overwrites it with the `client` accessor for the foreign key.

In this case, you may need to rename the association to something else, so that it doesn't collide with the foreign key. For example:

```ruby
class User
  include Mongoid::Document
  belongs_to :client_record, class_name: "Client", foreign_key: :client
end
```

It's not ideal, but working with legacy systems can be like that. :/

**Reply by CharlesPlantePulsar** [reply]
Thank you so much I didn't try to change the accessor instead !
